File: OOP/HW1/hw1.py
import re

number = 0
total = 0
pattern = re.compile("[^\d\+\-]")
while True:
    print("Please input your score:")
    line = input()
    result = pattern.search(line)
    if result is not None:
        print("illegal input, ending the program...", flush=True)
        quit(1)
    try:
        grade = float(line)
    except ValueError:
        grade = 0
        print("illegal input, ending the program...")
        quit(1)
    if grade == -1:
        break
    elif grade < 0 or grade > 100:
        print("illegal input, ending the program...", flush=True)
        quit(1)
    else:
        total += grade
        number += 1

average = total / number
print(f"The average is {average:.2f}.")

File: OOP/HW1/hw1.java
import java.util.Scanner;
import java.util.regex.*;
import java.text.DecimalFormat;

public class hw1 {
    private static final DecimalFormat df = new DecimalFormat("0.00");

    public static void main(String args[]) {
        int number; // The number of scores read.
        int sum; // The summation of all the scores read.
        int grade; // Holds each individual score as it is read from the keyboard.
        double average; // Holds the computed average.
        String input;
        String pattern = "[^\\d\\+\\-]";
        Pattern p = Pattern.compile(pattern);
        Matcher m;

        Scanner sc = new Scanner(System.in);
        number = 0;
        sum = 0;
        while (true) {
            System.out.println("Please input your score:");
            input = sc.nextLine();
            m = p.matcher(input);
            if (m.find()) {
                System.out.println("illegal input, ending the program...");
                System.exit(1);
            }
            try {
                grade = Integer.valueOf(input);
            } catch(NumberFormatException ex) {
                System.out.println("illegal input, ending the program...");
                grade = 0;
                System.exit(1);
            }
            if (grade == -1) {
                break;
            } else if (grade < 0 || grade > 100) {
                System.out.println("illegal input, ending the program...");
                System.exit(1);
            } else {
                sum += grade;
                ++number;
            }
        }
        average = (double) sum / number;
        System.out.println("The average is " + df.format(average) + ".");
        sc.close();
    }
}
